services/sn_analysis_service/app.py

```diff
@@ -0,0 +1,182 @@
+import os
+import functools
+import tempfile
+import io
+
+import numpy as np
+import matplotlib
+import arviz as az
+from quart import Quart
+from quart import request
+from astropy.table import Table
+import sncosmo
+
+from baselayer.log import make_log
+from baselayer.app.env import load_env
+
+_, cfg = load_env()
+log = make_log('sn_analysis_service')
+
+# we need to set the backend here to insure we
+# can render the plot headlessly
+matplotlib.use('Agg')
+rng = np.random.default_rng()
+
+app = Quart(__name__)
+app.debug = False
+
+default_analysis_parameters = {"source": "nugent-sn2p", "fix_z": False}
+
+
+def upload_analysis_results(results, data_dict):
+    """
+    Upload the results to the webhook.
+
+    TO BE COMPLETED
+    """
+    log("upload results to webhook")
+    assert isinstance(results, dict)
+    assert isinstance(data_dict, dict)
+    pass
+
+
+async def run_sn_model(dd):
+    """
+    Use `sncosmo` to fit data to a model with name `source_name`.
+
+    We expect the `inputs` dictionary to have the following keys:
+       - source: the name of the model to fit to the data
+       - fix_z: whether to fix the redshift
+       - photometry: the photometry to fit to the model (in csv format)
+       - redshift: the known redshift of the object
+
+    TODO in the next PR: send the results back to the webhook
+
+    """
+    data_dict = await dd
+
+    analysis_parameters = data_dict["inputs"].get("analysis_parameters", {})
+    analysis_parameters = {**default_analysis_parameters, **analysis_parameters}
+
+    source = analysis_parameters.get("source")
+    fix_z = analysis_parameters.get("fix_z") in [True, "True", "t", "true"]
+
+    # this example analysis service expects the photometry to be in
+    # a csv file (at data_dict["inputs"]["photometry"]) with the following columns
+    # - filter: the name of the bandpass
+    # - mjd: the modified Julian date of the observation
+    # - magsys: the mag system (e.g. ab) of the observations
+    # - flux: the flux of the observation
+    #
+    # the following code transforms these inputs from SkyPortal
+    # to the format expected by sncosmo.
+    #
+    results = {"status": "success", "message": "", "data": {}}
+    try:
+        data = Table.read(data_dict["inputs"]["photometry"], format='ascii.csv')
+        data.rename_column('mjd', 'time')
+        data.rename_column('filter', 'band')
+        data.rename_column('magsys', 'zpsys')
+
+        data['flux'].fill_value = 1e-6
+        data = data.filled()
+        data.sort("time")
+
+        redshift = Table.read(data_dict["inputs"]["redshift"], format='ascii.csv')
+        z = redshift['redshift'][0]
+    except Exception as e:
+        results.update(
+            {
+                "status": "failure",
+                "message": f"input data is not in the expected format {e}",
+            }
+        )
+        upload_analysis_results(results, data_dict)
```

Comment:
this function will be completed in the next PR, focused mainly on returning results